apis/placement/v1beta1/clusterresourceplacement_types.go

```diff
@@ -222,14 +222,141 @@ type PreferredClusterSelector struct {
 
 	// A cluster selector term, associated with the corresponding weight.
 	// +required
-	Preference ClusterSelectorTerm `json:"preference"`
+	Preference PreferredClusterSelectorTerm `json:"preference"`
+}
+
+// +enum
+type MetricSortOrder string
+
+const (
+	// Minimum instructs Fleet to sort in descending order, that is, the clusters with lower
+	// observed values are most preferred and should have higher weights.
+	//
+	// For example, with this order, if Fleet sorts all clusters by a specific metric where the
+	// observed values are in the range [10, 100], and a weight of 100 is specified;
+	// Fleet will assign:
+	//
+	// * a weight of 100 to the cluster with the minimum observed value (10); and
+	// * a weight of 0 to the cluster with the maximum observed value (100); and
+	// * a weight of 89 to the cluster with an observed value of 20.
+	//
+	//   It is calculated using the formula below:
+	//   (1 - ((20 - 10) / (100 - 10))) * 100 = 89
+	Minimum MetricSortOrder = "Minimum"
+
+	// Maximum instructs Fleet to sort in ascending order, that is, the clusters with higher
+	// observed values are most preferred and should have higher weights.
+	//
+	// For example, with this order, if Fleet sorts all clusters by a specific metric where
+	// the observed values are in the range [10, 100], and a weight of 100 is specified;
+	// Fleet will assign:
+	//
+	// * a weight of 0 to the cluster with the minimum observed value (10); and
+	// * a weight of 100 to the cluster with the maximum observed value (100); and
+	// * a weight of 11 to the cluster with an observed value of 20.
+	//
+	//   It is calculated using the formula below:
+	//   ((20 - 10)) / (100 - 10)) * 100 = 11
+	Maximum MetricSortOrder = "Maximum"
+)
+
+// MetricRange describes a required or preferred range for a metric.
+type MetricRange struct {
+	// Minimum is the minimum value of the range.
+	// If this value is not specified, the smallest observed value of the metric will
+	// be used.
+	// At least one of the minimum or maximum value must be specified.
+	// +optional
+	Minimum string `json:"minimum"`
+
+	// Maximum is the maximum value of the range.
+	// If this value is not specified, the largest observed value of the metric will be
+	// used.
+	// At least one of the minimum of maximum value must be specified.
+	// +optional
+	Maximum string `json:"maximum"`
+}
+
+// MetricMatcher is a specific metric requirement when picking clusters for resource placement.
+type MetricMatcher struct {
+	// Name is the name of the metric; it should be a Kubernetes label name.
+	// +required
+	Name string `json:"name"`
+
+	// Range are the required or preferred range for the metric.
+	//
+	// You may choose to specify multiple required/preferred ranges for a metric; these
+	// requirements will be OR'd.
+	//
+	// +required
+	Ranges []MetricRange `json:"range"`
```

Comment:
can we use operator style just like the lable selector? We just need to introduce something like

MetricsSelectorOpGe
MetricsSelectorOpLe
MetricsSelectorOpEq

and in the future, we can add MetricsSelectorOpIn, MetricsSelectorOpExist to support string based metrics

Reply:
Hi Ryan. The API has been updated to use expressions as prev. discussed.